File: c/vigenere.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

/*
 * Simple Vigenere cipher demonstration.
 * Usage: ./vigenere KEY text
 * Not intended for real cryptographic use.
 */

int main(int argc, char *argv[]) {
    if (argc < 3) {
        fprintf(stderr, "usage: %s key text\n", argv[0]);
        return 1;
    }
    char *key = argv[1];
    int key_len = strlen(key);
    int j = 0;
    for (int i = 2; i < argc; i++) {
        char *p = argv[i];
        if (i > 2) putchar(' ');
        for (; *p; p++) {
            if (isalpha((unsigned char)*p)) {
                char base = isupper((unsigned char)*p) ? 'A' : 'a';
                int shift = tolower(key[j % key_len]) - 'a';
                putchar(((*p - base + shift) % 26) + base);
                j++;
            } else {
                putchar(*p);
            }
        }
    }
    putchar('\n');
    return 0;
}

File: py/vigenere.py
import sys

def vigenere(text, key):
    res = []
    j = 0
    klen = len(key)
    for ch in text:
        if ch.isalpha():
            base = ord('A') if ch.isupper() else ord('a')
            shift = ord(key[j % klen].lower()) - ord('a')
            res.append(chr((ord(ch) - base + shift) % 26 + base))
            j += 1
        else:
            res.append(ch)
    return ''.join(res)

if __name__ == '__main__':
    if len(sys.argv) < 3:
        print('usage: python vigenere.py key text')
        sys.exit(1)
    key = sys.argv[1]
    text = ' '.join(sys.argv[2:])
    print(vigenere(text, key))
